pkg/sanity/resources.go

```diff
@@ -222,107 +211,119 @@ func (cl *Resources) registerSnapshot(offset int, id string) {
 
 func (cl *Resources) registerSnapshotNoLock(offset int, id string) {
 	ExpectWithOffset(offset, id).NotTo(BeEmpty(), "ID for register snapshot is missing")
-	if cl.snapshots == nil {
-		cl.snapshots = make(map[string]bool)
-	}
-	cl.snapshots[id] = true
+	Debugf("registering snapshot resource %s", id)
+	cl.managedResourceInfos = append(cl.managedResourceInfos, resourceInfo{
+		id:   id,
+		data: snapshotInfo{},
+	})
 }
 
-func (cl *Resources) unregisterSnapshot(offset int, id string) {
+func (cl *Resources) unregisterResource(offset int, id string) {
 	cl.mutex.Lock()
 	defer cl.mutex.Unlock()
-	cl.unregisterSnapshotNoLock(offset+1, id)
+	cl.unregisterResourceNoLock(offset+1, id)
 }
 
-func (cl *Resources) unregisterSnapshotNoLock(offset int, id string) {
-	ExpectWithOffset(offset, id).NotTo(BeEmpty(), "ID for unregister snapshot is missing")
-	if cl.snapshots != nil {
-		delete(cl.snapshots, id)
+func (cl *Resources) unregisterResourceNoLock(offset int, id string) {
+	ExpectWithOffset(offset, id).NotTo(BeEmpty(), "ID for unregister resource is missing")
+	// Find resource info with the given ID and remove it.
+	for i, resInfo := range cl.managedResourceInfos {
+		if resInfo.id == id {
+			Debugf("unregistering resource %s", id)
+			cl.managedResourceInfos = append(cl.managedResourceInfos[:i], cl.managedResourceInfos[i+1:]...)
+			return
+		}
 	}
 }
 
-// Cleanup calls unpublish methods as needed and deletes all volumes and
-// snapshots.
+// Cleanup calls unpublish methods as needed and deletes all managed resources.
 func (cl *Resources) Cleanup() {
 	cl.mutex.Lock()
 	defer cl.mutex.Unlock()
 	ctx := context.Background()
 
-	cl.deleteSnapshots(ctx, 2)
-	cl.deleteVolumes(ctx, 2)
+	// Clean up resources in LIFO order to account for dependency order.
```

Comment:
PVC and VolumeSnapshot have independent life cycles in Kubernetes.  If we fix the sanity test to make sure snapshots are always deleted before volumes are deleted, driver maintainers won't know it is possible for users to delete PVCs first and may run into the problems later on in production.

So I think we should actually have sanity tests that delete snapshot before and after the volume's deletion to make sure a driver can handle this independent life cycle.  Yes, this means driver will have to implement logic to do soft deletion if volume and snapshot are dependent to each other on the storage system in order to pass the sanity tests.

Reply:
I had commented on https://github.com/kubernetes-csi/csi-test/pull/289
Spec provides flexibility for PVC and its snapshots to be dependent or not. Spec https://github.com/container-storage-interface/spec/blob/master/spec.md says following:

===
CSI plugins SHOULD treat volumes independent from their snapshots.

If the Controller Plugin supports deleting a volume without affecting its existing snapshots, then these snapshots MUST still be fully operational and acceptable as sources for new volumes as well as appear on ListSnapshot calls once the volume has been deleted.

When a Controller Plugin does not support deleting a volume without affecting its existing snapshots, then the volume MUST NOT be altered in any way by the request and the operation must return the **FAILED_PRECONDITION** error code and MAY include meaningful human-readable information in the status.message field.
====

So, the test case should not validate just one behavior as the spec allows volumes and snapshots to be dependent and that is the case for some of the drivers.

Reply:
With LIFO-ordered deletion implemented by this PR, I _think_ we should satisfy the minimum requirement that the spec mandates, which is that plugins do not or cannot treat volumes and snapshots independently. That is, if a volume is created from a snapshot, we first delete the snapshot and then the volume. (I also assume the same logic should apply if a volume is sourced from another volume.)

However, @apurv15 mentioned in [a previous comment](https://github.com/kubernetes-csi/csi-test/pull/297#issuecomment-705973373) that csi-test built from this change would cause more tests to fail. 🤔 I guess I should give the PR another hard to look to see if I missed anything. (@apurv15, did you ever get around running it with the enhanced debug log messages as I suggested?)

@xing-yang would it make sense to file a separate issue to track creating a new test for the missed scenario you mentioned (while taking into account the specifics of the spec as @apurv15 quoted above)? IMHO, this PR should focus on addressing the regression I introduced and returning to the [original cleanup order](https://github.com/kubernetes-csi/csi-test/blob/4b9faee268d03e35c94499387aeb4cdb532f4e58/pkg/sanity/controller.go#L699-L712).

Reply:
There are also storage systems that cannot delete snapshots before deleting PVCs. Our existing e2e tests also delete in a specific order, deleting snapshots first and then deleting PVCs. However, that caused some CSI drivers to fail because they require an opposite ordering of deletion which is PVCs must be deleted before snapshots are deleted.  There's an issue opened to track and fix this in e2e tests so that deletion of PVC and snapshots should be independent of each other:
https://github.com/kubernetes/kubernetes/issues/96719

Reply:
@xing-yang how should csi-test (and the e2e tests) account for the two possible orderings?

Should we try one order, and if we receive a particular error, try the inverse?

Or is the expectation that the driver acknowledges delete requests and completes the cleanup asynchronously once both the snapshot and volume deletes have been issued?

csi-test currently does not check that deletes have completed, so we might naturally support the latter scenario but not the former.

Reply:
We should test both cases:
- Delete volumes before deleting snapshots
- Delete snapshots before deleting volumes

CSI driver should pass both tests.  This means CSI driver will need to keep reference count internally and allow deletion to happen successfully regardless of the delete ordering.

Reply:
CC @msau42

Reply:
This doesn't need to be solved in this PR. It's more like an idea for some dedicated tests.

Reply:
Regarding cleanup order: of both cases are valid, then picking one for cleanup (LIFO at the moment) is correct, i.e. we can merge this PR.

Reply:
Regarding this comment `Clean up resources in LIFO order to account for dependency order.`, we should remove the text "to account to dependency order" as we are not trying to make sure snapshots are deleted before volumes are deleted.

Reply:
In its current form, the code cleans up resources "backwards": the resource (either volume or snapshot) that was created last is cleaned up first, followed by cleaning up the resource that was created second-to-last, and so on. Given that sourcing volumes must exist before the snapshots they are created from, this means that we first delete a snapshot followed by its sourcing volume.

Doesn't that imply that we _do_ make sure that snapshots are deleted before volumes are @xing-yang?

Or are we talking about the scenario where a volume is being created from a snapshot and a possible dependency in that chain?

Reply:
We discussed this a bit further on Monday. @msau42 pointed out that it is valid when a CSI driver refuses to delete PVCs that have snapshots. It is *also* valid to refuse deleting snapshots when there is still a PVC. In other words, a CSI driver only needs to support *one* order of deletions, but not both.

This means that picking LIFO order will work for some drivers, but not all.

The right approach is to retry deletion after it failed once, i.e. try to delete snapshot, delete pvc, then if snapshot deletion failed, try again. I think this can be added relatively easily by iterating twice over all resources instead of just once.

We also discussed writing tests that explicitly cover both deletion orders. One test can do "delete snapshot, delete pvc, delete snapshot again (if necessary)", the other "delete pvc, delete snapshot, delete pvc again (if necessary)". This then works with all drivers and ensures that both is tested. But this should be added in a separate PR.